src/proto/debug/mod.rs

```diff
@@ -9,24 +9,134 @@
+pub use self::exception::ExceptionType;
+
+mod context;
+mod exception;
 
 /// The debugging support protocol allows debuggers to connect to a UEFI machine.
+/// It is expected that there will typically be two instances of the EFI Debug Support protocol in the system.
+/// One associated with the native processor instruction set (IA-32, x64, ARM, RISC-V, or Itanium processor
+/// family), and one for the EFI virtual machine that implements EFI byte code (EBC).
+/// While multiple instances of the EFI Debug Support protocol are expected, there must never be more than
+/// one for any given instruction set.
+///
+/// NOTE: OVMF only implements this protocol interface for the virtual EBC processor
 #[repr(C)]
 #[unsafe_guid("2755590c-6f3c-42fa-9ea4-a3ba543cda25")]
 #[derive(Protocol)]
 pub struct DebugSupport {
     isa: ProcessorArch,
-    // FIXME: Add the mising parts of the interface. Beware that it features
-    //        unsafety in the form of an unchecked processor index.
+    get_maximum_processor_index:
+        extern "efiapi" fn(this: &mut DebugSupport, max_processor_index: &mut usize) -> Status,
+    register_periodic_callback: unsafe extern "efiapi" fn(
+        this: &mut DebugSupport,
+        processor_index: usize,
+        periodic_callback: Option<unsafe extern "efiapi" fn(SystemContext)>,
+    ) -> Status,
+    register_exception_callback: unsafe extern "efiapi" fn(
+        this: &mut DebugSupport,
+        processor_index: usize,
+        exception_callback: Option<unsafe extern "efiapi" fn(ExceptionType, SystemContext)>,
+        exception_type: ExceptionType,
+    ) -> Status,
+    invalidate_instruction_cache: unsafe extern "efiapi" fn(
+        this: &mut DebugSupport,
+        processor_index: usize,
+        start: *mut c_void,
+        length: u64,
+    ) -> Status,
 }
 
 impl DebugSupport {
     /// Returns the processor architecture of the running CPU.
     pub fn arch(&self) -> ProcessorArch {
         self.isa
     }
+
+    /// Returns the maximum value that may be used for the processor_index parameter in
+    /// `register_periodic_callback()` and `register_exception_callback()`.
+    ///
+    /// Note: Applications built with EDK2 (such as OVMF) always return `0` as of 2021-09-15
+    pub fn get_maximum_processor_index(&mut self) -> usize {
+        // initially set to a canary value for testing purposes
+        let mut max_processor_index: usize = usize::MAX;
+
+        // per the UEFI spec, this call should only return EFI_SUCCESS
+        let _ = (self.get_maximum_processor_index)(self, &mut max_processor_index);
+
+        max_processor_index
+    }
+
+    /// Registers a function to be called back periodically in interrupt context.
+    /// Pass `None` for `callback` to deregister the currently registered function for
+    /// a specified `processor_index`. Will return `Status::INVALID_PARAMETER` if
+    /// `processor_index` exceeds the current maximum from `Self::get_maximum_processor_index`.
+    ///
```

Comment:
Might be good to mention this from the spec:

> In addition, no portion of the debug agent that runs in interrupt context may make any calls to EFI services or other protocol interfaces.

So if I'm understanding that right, it would rule out things like using the [global system table pointer](https://docs.rs/uefi-services/0.9.0/uefi_services/fn.system_table.html) from uefi-services, or calling the global logger from within the callback?

Reply:
Added! That's my understanding as well, it seems to be strictly for exposing the system context and/or processor exceptions to the host-side debug agent. Maybe these functions should also go back to being `unsafe`?

Reply:
Yeah I think making them unsafe is probably the right call.